hub/core/tensor.py

```diff
@@ -49,41 +50,48 @@ def create_tensor(key: str, storage: StorageProvider, meta: dict):
     if tensor_exists(key, storage):
         raise TensorAlreadyExistsError(key)
 
-    validate_tensor_meta(meta)
-
-    write_tensor_meta(key, storage, meta)
-    write_index_map(key, storage, [])
+    TensorMeta.create(key, storage, htype=htype, **kwargs)
+    IndexMeta.create(key, storage)
 
 
 def add_samples_to_tensor(
     array: np.ndarray,
     key: str,
     storage: StorageProvider,
     batched: bool = False,
+    tensor_meta: TensorMeta = None,
+    index_meta: IndexMeta = None,
 ):
     """Adds samples to a tensor that already exists. `array` is chunked and sent to `storage`.
     For more on chunking, see the `generate_chunks` method.
+
     Args:
         array (np.ndarray): Array to be chunked/written. Batch axis (`array.shape[0]`) is optional, if `array` does
-        have a batch axis, you should pass the argument `batched=True`.
-        key (str): Key for where the chunks, index_map, and meta will be located in `storage` relative to it's root.
-        storage (StorageProvider): StorageProvider for storing the chunks, index_map, and meta.
+            have a batch axis, you should pass the argument `batched=True`.
+        key (str): Key for where the chunks, index_meta, and meta will be located in `storage` relative to it's root.
+        storage (StorageProvider): StorageProvider for storing the chunks, index_meta, and meta.
         batched (bool): If True, the provided `array`'s first axis (`shape[0]`) will be considered it's batch axis.
-        If False, a new axis will be created with a size of 1 (`array.shape[0] == 1`). default=False
+            If False, a new axis will be created with a size of 1 (`array.shape[0] == 1`). default=False
+        tensor_meta (TensorMeta): Optionally provide a `TensorMeta`. If not provided, it will be loaded from `storage`.
+        index_meta (IndexMeta): Optionally proivide an `IndexMeta`. If not provided, it will be loaded from `storage`.
+
     Raises:
         TensorDoesNotExistError: If a tensor at `key` does not exist. A tensor must be created first using
-        `create_tensor(...)`.
+            `create_tensor(...)`.
     """
 
     if not tensor_exists(key, storage):
         raise TensorDoesNotExistError(key)
 
-    index_map = read_index_map(key, storage)
-    tensor_meta = read_tensor_meta(key, storage)
+    if index_meta is None:
```

Comment:
Will this throw error if there is no IndexMeta? Or this will be caught somewhere else?

If this can be None or not exist, should we throw error here?

Reply:
yep! in the function `tensor_exists(...)`, it checks if both an `index_meta.json` and `tensor_meta.json` exists at the provided key

Reply:
it then throws the `TensorDoesNotExistError`

Reply:
nit-pick: the None check isn't needed here actually and can result in weird errors when the code above this doesn't catch it. this could have been an assert instead.